crates/project-model/src/project_json.rs

```diff
@@ -202,6 +243,8 @@ struct CrateData {
     is_proc_macro: bool,
     #[serde(default)]
     repository: Option<String>,
+    #[serde(default)]
+    target_spec: Option<TargetSpecData>,
```

Comment:
Some notes on JSON side of things here:

I'd maybe call the top-level thing `build_info`, as thre might be multiple targets (in buck2 sense) concerning a single crate. And, well, target is already overloaded, so it's better to avoid pilling more meanings onto that poor word.

I _think_ the core feature here should be an inclusion of opaque blob of data, which is transmitted over LSP so that the custom client may interpret this. So, I'd say the basis should look like

```
build_info: Option<BuildInfo>

struct BuildInfo {
    data: serde_json::Value
}
```

Eg, one can imagine a situation where _all_ builds are remote, and the way you "run" software is by issueing an HTTP request to a remote execution server, and opening a page in the web-browser with results. In this situation, the `data` would be an URL to call, and the client-side extension would take care of HTTPing.

But we also should support "standard" workflows where you just run some command line. 

So, something like (pardon my TypeScript)

```
struct BuildInfo {
    data: serde_json::Value
    runnables: Vec<Runnable>
}

struct Runnable {
    kind: "build" | "check" | "test" | "run" | string,
    label: string,
    argv: string[],
    cwd: string
}
```

This _shouldn't_ use a json map, but rather an array, to be more flexible. Eg, in the future we might want to add `--release` and non-`--release` versions of tests, and with a list it should be simpler. More generally, I've recently read some JSON API guidlines somewhere (don't remember whether though), and one guideline that stuck with me is that you should represent maps in JSON as arrays (assoc lists). I think that's a good guideline and something we should be using in gerneral. 

_Then_, the `kind` should be a semi-open enum. rust-analyzer _should_ treat `build`, `run`, `check` and `test` specially, but it should also be able to show other tasks in GUI. Eg, you can imagine a "flush" task for an embedded build system, or some such. 

We _might_ want to split `check` into two --- one check that is run for the user and displays errors in the terminal, and another check that runs with message-format=json and is used for flycheck. I guess we could call those `check` and `flycheck`?

For _tests_, we need to think how to enable filtering. The core feature is that you can run a single test using `cargo test -- $TEST_NAME`, and all tests using `cargo test`. Note that this might be very different in non-cargo build systems. Eg, it could be `builder test` and `builder test --test-filter=$TEST_NAME`, where `builder test --test-filter=''` would run zero tests, rather than all tests. So I think for test, we should also split it into two: "testall" and "testone". And, for "testone", I _think_ we want to make `argv` a template. That is, argv could look like

```
["builder", "test", "--test-filter=$$TEST_NAME$$", "./buildfile.b"]
```

We shouldn't assume that filter is always last, and that its always a separate argument.

Reply:
Thanks for the feedback here!

> I'd maybe call the top-level thing `build_info`, as thre might be multiple targets (in buck2 sense) concerning a single crate. And, well, target is already overloaded, so it's better to avoid pilling more meanings onto that poor word.

That's a fair point. I didn't intend for the naming to be Buck/Bazel-esque "targets", I mean to invoke the [target](https://doc.rust-lang.org/cargo/reference/cargo-targets.html) concept in _Cargo_, of all things. In any case, `target` is too overloaded and I'll move over to `build_info`.

> I think the core feature here should be an inclusion of opaque blob of data, which is transmitted over LSP so that the custom client may interpret this.

Is the idea that a companion extension to `rust-lang.rust-analyzer` (e.g., something for Buck or Bazel) would be responsible for assembling that opaque blob of data into a some sort of meaningful runnable command or is more that since I'm making this change now, I should make it more futureproof ("[B]ut we also should support "standard" workflows where you just run some command line.")? I ask since I'm working on a variable interpreter/template thingy, which—I think—should make expressing arbitrary build/test commands a lot easier in rust-analyzer itself, which reduces the pressure/need for end-users to install other extensions for runnables.

> This shouldn't use a json map, but rather an array, to be more flexible. Eg, in the future we might want to add `--release` and non-`--release` versions of tests, and with a list it should be simpler. More generally, I've recently read some JSON API guidlines somewhere (don't remember whether though), and one guideline that stuck with me is that you should represent maps in JSON as arrays (assoc lists). I think that's a good guideline and something we should be using in gerneral.

That seems reasonable to me! I was considering how to tackle this for Buck[^1] and this seems like a good approach. 

> Then, the `kind` should be a semi-open enum. rust-analyzer should treat `build`, `run`, `check` and `test` specially, but it should also be able to show other tasks in GUI. Eg, you can imagine a "flush" task for an embedded build system, or some such. 

Yeah, that seems smart.

> We might want to split `check` into two --- one `check` that is run for the user and displays errors in the terminal, and another check that runs with `message-format=json` and is used for flycheck. I guess we could call those `check` and `flycheck`?

I think now's a good time to make this change, especially if the new runnable API is a breaking change.

For tests, we need to think how to enable filtering. The core feature is that you can run a single test using `cargo test -- $TEST_NAME`, and all tests using `cargo test`. Note that this might be very different in non-cargo build systems. Eg, it could be `builder test` and `builder test --test-filter=$TEST_NAME`, where `builder test --test-filter=''` would run zero tests, rather than all tests. So I think for test, we should also split it into two: `"testall"` and `"testone"`. And, for `"testone"`, I _think_ we want to make `argv` a template. That is, argv could look like: [redacted for quoting ease]

I don't think I've fully gotten this into my head yet (sorry, lots of travel!) but I think I see the utility of a change like this.

I think I'd like like a `debug` variant as well, but I haven't really through _this specific aspect_ through.

[^1]: For context, buck2 is moving to a [modifier-style approach](https://github.com/facebook/buck2/blob/main/docs/rfcs/drafts/cfg-modifiers/api.md#cli-modifier). A user can write `buck2 build repo//foo:bar?linux,release`, which translate to "`build repo//foo:bar` for Linux using `release`". there's a more verbose `--modifier`-style API, but I suspect that both options can work when called from rust-analyzer _or_ I suspect the Buck team will be flexible in implementing this.